### PR TITLE
Add support for null conditionals in implicit expressions.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Added case in ImplicitExpression handling to understand question marks.
- text?. is special compared to text. because with text. we currently validate content after text. to determine if it's an expression or if it's a period. Now with ?. we always treat it as an expression because ?. is not a useful sentance of any kind.
- Added unit tests to validate new implicit expression handling
- Added runtime and design time code generation tests to validate null propagation.

#44

*Note: Will need to evaluate the editing experience in the editor once VS bits become available due to the complex nature of implicit expressions.*

### CODE REVIEW

File: test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs [hunk 1]
```diff
@@ -17,6 +18,105 @@ public override ParserBase CreateCodeParser()
...
+                    { "foo?!", "foo" },
+                    { "foo?.", "foo?." },
+                    { "foo??.", "foo" },
+                    { "foo?.(abc)", "foo?." },
+                    { "foo?.<p>", "foo?." },
+                    { "foo?.bar", "foo?.bar" },
+                    { "foo?.bar<p>", "foo?.bar" },
+                    { "foo??.bar<p>", "foo" },
+                    { "foo?.bar(false)?.<p>", "foo?.bar(false)?." },
+                    { "foo?.bar(false)?.baz", "foo?.bar(false)?.baz" },
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs]: - mix some tests like `foo??bar` in here
- avoid "foo", "bar", et cetera when adding new tests

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullPropagationExpressions.cshtml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿@{
+    @ViewBag?.Data
+    @ViewBag.Method?(Hello World)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullPropagationExpressions.cshtml]: Has support for `Method?(...)` been added to the language recently?  Null-conditional syntax previously included only properties (`?.Property`) and indexers (`?[index]`).

Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullPropagationExpressions.cshtml]: self-nit: `Property?.Method()` also works :smiley:

File: test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs [hunk 3]
```diff
@@ -17,6 +18,105 @@ public override ParserBase CreateCodeParser()
...
+                    { "foo?(", "foo?(", AcceptedCharacters.Any, missingEndParenError(5) },
+                    { "foo?(bar", "foo?(bar", AcceptedCharacters.Any, missingEndParenError(5) },
+                    { "foo?(abc)", "foo?(abc)", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?(<p>", "foo?(", AcceptedCharacters.Any, missingEndParenError(5) },
+                    { "foo?(bar.<p>", "foo?(bar.", AcceptedCharacters.Any, missingEndParenError(5) },
+                    { "foo??(bar<p>", "foo", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?(abc)?", "foo?(abc)", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?(abc)?(def", "foo?(abc)?(def", AcceptedCharacters.Any, missingEndParenError(11) },
+                    { "foo?(abc)?(def)", "foo?(abc)?(def)", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?(abc)?.bar?(def)", "foo?(abc)?.bar?(def)", AcceptedCharacters.NonWhiteSpace, noErrors },
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs]: Doubt any of the above are valid C#.

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs [hunk 4]
```diff
@@ -388,7 +388,33 @@ private bool MethodCallOrArrayIndex(AcceptedCharacters acceptedCharacters)
...
+
+                    if (next != null)
+                    {
+                        if (next.Type == CSharpSymbolType.Dot)
+                        {
+                            // Accept null propagation dot operator (?.).
+                            AcceptAndMoveNext();
+                            AcceptAndMoveNext();
+
+                            // If the next piece after the ?. is a keyword or identifier then we want to continue.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs]: What keyword should work in this context?  E.g. `Value?.nameof(...)` isn't valid.

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs [hunk 5]
```diff
@@ -388,7 +388,33 @@ private bool MethodCallOrArrayIndex(AcceptedCharacters acceptedCharacters)
...
+                            return At(CSharpSymbolType.Identifier) || At(CSharpSymbolType.Keyword);
+                        }
+                        else if (next.Type == CSharpSymbolType.LeftParenthesis ||
+                            next.Type == CSharpSymbolType.LeftBracket)
+                        {
+                            // It's either ?[ or ?(, accept the question mark because it's a null propagation operator.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs]: Again, what does `?(` mean?

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs [hunk 6]
```diff
@@ -388,7 +388,33 @@ private bool MethodCallOrArrayIndex(AcceptedCharacters acceptedCharacters)
...
-                if (CurrentSymbol.Type == CSharpSymbolType.Dot)
+                if (At(CSharpSymbolType.QuestionMark))
+                {
+                    var next = Lookahead(count: 1);
+
+                    if (next != null)
+                    {
+                        if (next.Type == CSharpSymbolType.Dot)
+                        {
+                            // Accept null propagation dot operator (?.).
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs]: This feature is now consistently called "Null-conditional operators".  See https://github.com/dotnet/roslyn/wiki/New-Language-Features-in-C%23-6

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs]: https://github.com/dotnet/roslyn/wiki/New-Language-Features-in-C%23-6#null-conditional-operators specifically says
> Note that an invocation (a parenthesized argument list) cannot immediately follow the ? operator – that would lead to too many syntactic ambiguities.

The workaround for delegates is a null-conditional `Invoke` call e.g. `delegate?.Invoke(...)`

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullPropagationExpressions.cshtml (hunk 2 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullPropagationExpressions.cshtml]: Oh man I thought that Method?() was supported for delegates. Just tried it out, I be wrong, will remove.

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs]: How is that not valid? If there's a property which happens to be a delegate named `nameof` it works fine. Same for `class`, `typeof`, `for` etc.

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.cs]: https://github.com/aspnet/Razor/pull/396#discussion_r30571892

**Comment by NTaylorMullen** [comment]
Updated.

File: test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs [hunk 7]
```diff
@@ -17,6 +18,96 @@ public override ParserBase CreateCodeParser()
...
+        public static TheoryData NullConditionalOperatorData_Dot
+        {
+            get
+            {
+                // implicitExpression, expectedImplicitExpression
+                return new TheoryData<string, string>
+                {
+                    { "val?", "val" },
+                    { "val??", "val" },
+                    { "val??more", "val" },
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs]: Not something that should change in this PR but this case is interesting. Does the test mean null coalescing expressions aren't supported in implicit expressions, at least outside brackets or parentheses? Might be worth a bug since this case is unlikely to be an typo e.g. would expect whitespace after `??` or `???` if this were English.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/NullConditionalExpressions.DesignTime.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+#line hidden
+#line 4 "NullConditionalExpressions.cshtml"
+                               
+    
+
+#line default
+#line hidden
+
+#line 5 "NullConditionalExpressions.cshtml"
+__o = ViewBag?.Method(Value?[23]?.More)(SecondValue)?["key"];
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/NullConditionalExpressions.DesignTime.cs]: Not something that should change in this PR but might be worth filing a bug to remove this corner case from implicit expression support. Adjacent parenthesized expressions happen only when you have a `Func<..., Action<...>>`, `Func<..., Func<...>>` or similar nested delegates and you're calling the whole thing in one expression. Seems more likely the expression ends after the first method invocation and the next part begins a new parenthesized sentence.

Fix relative to this PR's version of the files would be at https://github.com/aspnet/Razor/pull/396/files#diff-d2cd6cc81ade624cba92cccb004b6afdR389

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullConditionalExpressions.cshtml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    @ViewBag?.Data
+    @ViewBag.IntIndexer?[0]
+    @ViewBag.StrIndexer?["key"]
+    @ViewBag?.Method(Value?[23]?.More)(SecondValue)?["key"]
+}
+
+@ViewBag?.Data
+@ViewBag.IntIndexer?[0]
+@ViewBag.StrIndexer?["key"]
+@ViewBag?.Method(Value?[23]?.More)(SecondValue)?["key"]
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullConditionalExpressions.cshtml]: Why bother testing nested delegates to this extent? It's a corner case that should be visible in at most one or two low-level (`CSharpImplicitExpressionTest`) cases.

**Comment by dougbu** [comment]
> What keyword should work in this context? E.g. Value?.nameof(...) isn't valid.

> How is that not valid? If there's a property which happens to be a delegate named nameof it works fine. Same for class, typeof, for etc.

`nameof` was a bad example in my comment. You can define a delegate or other property / field named `nameof` or using most other contextual keywords. However the same is not true of `class` or `typeof` unless used with in a C# verbatim (`@`) identifier.

Unfortunately I now see the Razor list of C# keywords (in `CSharpKeywordDetector`) mixes contextual keywords in with `class` and other non-contextual keywords. Carry on :frowning:

**Comment by dougbu** [comment]
:shipit: after minimizing testing of the extreme corner case mentioned above.

File: test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs (hunk 7 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs]: Nope, not supported. If you want to do pieces of code that typically have spaces the route to take is `@(...)`. Same reason why `@a+1` doesn't work :smile:

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullConditionalExpressions.cshtml (hunk 9 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/NullConditionalExpressions.cshtml]: This is just combining everything in one, will remove the nested delegate part of this.
